[영화 극장 선택] 오둥이 3, 4단계 미션 제출합니다.

murjune opened the pull request:
안녕하세요 말리빈! 드디어 미션을 제출하게 되었네요  

1,2 단계에 comment 남겨주신 부분에 답글을 달아 두었습니다 !!  
이번에도 잘 부탁드립니다 🙇‍♂️  

https://github.com/woowacourse/android-movie-theater/pull/62#discussion_r1590740747 
https://github.com/woowacourse/android-movie-theater/pull/62#discussion_r1597470188  

## 푸시 알림 및 권한 요청 영상

1) home 에서 권한 수락 후, 예매 하기

https://github.com/woowacourse/android-movie-theater/assets/87055456/e47c5318-ce47-4087-bc45-8313f9ee5a53

2) home 에서 권한 거절 후, setting 에서 권한 수락 후, 예매 하기

https://github.com/woowacourse/android-movie-theater/assets/87055456/65ee6c7c-a48f-42c0-896a-cb42a40daa5d


3) Android 12 핸드폰 (권한 체크 X, setting에 자동으로 푸시 알림 받기 active)

https://github.com/woowacourse/android-movie-theater/assets/87055456/ed35dacd-1583-468a-b78c-c0de6f2dc0aa

## 궁금한 점 & 질문 사항

### 1. Espresso 테스트 에서 dataBinding 객체 가져오기
인수 테스트에서 Binding 객체를 가져오려 노력했지만, 어떻게 가져와야할지 모르겠습니다..  
아래는 제가 시도해봤던 코드입니다.  

```kotlin
 private lateinit var binding: FragmentMovieListBinding
 
launchFragmentInContainer<MovieListFragment>().onFragment {
    binding = FragmentMovieListBinding.bind(it.requireView()) 
}

// 이와 같이 테스트 코드를 작성하고 싶었습니다
onView(withId(binding.rvMovies.id)).perform(...)
```  

 새로운 Binding 객체를 생성하고 있기에 에러가  발생한 것 같습니다..
![image](https://github.com/android/architecture-components-samples/assets/87055456/315cad6d-975b-4273-8d86-f31c49ba2066)

## 2. Permission 요청 로직

Home 에서 권한 요청을 하는 것이  미션 요구 사항이였으나  
정말 필요할 때 권한을 요청하는 것이 사용자 친화적인 UX 라고 생각했습니다.
따라서, 다음과 같이 퍼미션을 요청 로직으로 구현했습니다.
```
- 푸시 알림 퍼미션 로직
- Android Tiramisu(13) 이상에서만 동작
    1) Home - 맨 처음에만 퍼미션 체크를 함
    2) Setting
    - 1번째 요청 - ActivityResultLauncher 에서 권한 요청
    - 2번째 요청 
      - Explanation Dialog show -> ActivityResultLauncher 에서 권한 요청
    - 3번째 요청부터는 아무 것도 하지 않음
- Android Tiramisu(13) 미만에서는 자동으로 권한이 부여되어 있도록 함
```

이때, 이미 2번 거절을 한 경우에는 `안드로이드 시스템 Setting` 화면으로 이동시키거나 이동 확인 Dialog 를 보여주도록 구현하고 싶었으나
2번 거절을 한 경우를 구분하는 방법을 찾지 못했습니다.
혹시, 이에 대한 좋은 방법이나 힌트를 주실 수 있을까요??
```kotlin
 Intent().also { intent ->
        intent.action = Settings.ACTION_APP_NOTIFICATION_SETTINGS
        intent.putExtra(Settings.EXTRA_APP_PACKAGE, context.packageName)
        startActivity(intent)
    }
```

Hevton committed:
[영화 극장 선택] 팡태 1, 2단계 미션 제출합니다 (#66)

murjune committed:
test: Mockk study

murjune committed:
:truck: package 구조 정리

murjune committed:
docs: 3,4 단계 기능명세서 작성

murjune committed:
:recycle: item_movie_list ui 수정

murjune committed:
🐛  MovieList package fix

murjune committed:
✨  Movie DB 만들기

  - [x]: Movie DB 만들기
  - [x]: Reservation Table(=Entity) 만들기
  - [x]: Seat Table(=Entity) 만들기
  - [x]: dao
  - [x]: dao Test

murjune committed:
✨  MovieMapper, Reservation model 생성

murjune committed:
✨  MovieRepository

murjune committed:
✨  TheaterSeatActivity DB 연결

- [x]: Mapper (Entity -> Domain Model)
 - [x]: Movie Repository
- [x]: 좌석 선택 뷰에서 예매 확인 시 Room 에 저장 하고 return 받은 id를 넘겨 주자
  - [x]: 저장은 worker Thread, startActivity 는 ui Thread 에서 하자고~
 - [x]: 기존 객체 직렬화 방식 -> id 넘겨 주는 방식으로 변경

murjune committed:
✨  예매 내역 리사이클러 뷰 만들기

  - [x]: Adapter
  - [x]: ItemView xml
  - [x]: ViewHolder

murjune committed:
✨ NotificationDataStore

-  [x] SharedPref 에 푸시 알림 여부를 저장 한다.

murjune committed:
임시 저장

murjune committed:
✨permission 분기 처리

- 푸시 알림 퍼미션 로직
-
    1) Home - 맨 처음에만 퍼미션 체크를 함
-
    2) Setting

    - 1번째 요청 아무 것도 없음
    - 2번째 요청에만 Explanation Dialog 보여줌
    - 3번째 요청부터도 아무 것도 없음 (이게.. 1, 3 번째 구분하는게 모르겠음)

Commit:
:white_check_mark: Presenter 테스트 구현

murjune committed:
:truck: ReservationContract 패키지 분리

murjune committed:
:recycle: reservation id 수정

- rv_movies -> rv_reservation_movies

murjune committed:
:sparkles: MovieDetailActivityTest

murjune committed:
:recycle: 함수 분리

murjune committed:
:recycle: 30분 전 푸시 알림, custom icon 적용

murjune committed:
:bug: MovieListFragmentTest

- 멀티 뷰타입으로 바뀐 다음 테스트 깨지는거 수정

murjune committed:
:sparkles: 32 일 때는 preference canNotify 를 true, 33 이상은 false 로 변경

murjune committed:
:truck: PurchaseConfirmation packaging

murjune committed:
:sparkles: MovieRepositoryStore 구현

- 동적으로 MovieRepository를 주입하기 위해 구현

murjune committed:
:sparkles: PurchaseConfirmationActivityTest 개선

murjune committed:
:recycle: ktLintFormat

murjune committed:
docs: README 수정

질문 사항

murjune committed:
:recycle: 리뷰 반영

malibinYun commented on app/src/androidTest/java/woowacourse/movie/data/dao/MovieDaoTest.kt:
```diff
@@ -0,0 +1,83 @@
+    @Test
+    @DisplayName("예매 내역을 저장 하고, id로 영화 내역을 조회할 수 있다.")
+    fun insert_and_load_test() {
+        repeat(1000) {
+            // given
```
1000번의 횟수가 정말로 유의미한 결과를 가지는 테스트인가요?

murjune replied:
앗..! 삭제하는 것을 실수했네요..  
처음에 스레드 없이도 잘 돌아가는지 확인용으로 작성한 코드였습니다.

malibinYun commented on app/src/androidTest/java/woowacourse/movie/util/ContextUtils.kt:
```diff
@@ -0,0 +1,6 @@
+
+import android.content.Context
+import androidx.test.core.app.ApplicationProvider
+
+val context: Context get() = ApplicationProvider.getApplicationContext<Context>()
```
테스트에서 써야할 흐름과 맞지 않는 context를 잘못 접근하게될 가능성이 보이네요 :)
예를들어, 특정 Activity의 Context를 접근하려 특정 변수의 context를 접근하려했는데, 실수로 글로벌한 함수의 context에 접근해서 에러가 발생한다면, 디버깅하기 굉장히 곤란해질 것으로 예상되네요 :)

murjune replied:
`testApplicationContext` 로 네이밍 수정하겠습니다!

malibinYun commented on app/src/main/java/woowacourse/movie/data/DefaultMovieRepository.kt:
```diff
@@ -0,0 +1,31 @@
+    private val movieDao: MovieDao,
+) : MovieRepository {
+    override fun loadReservedMovies(): Result<List<Reservation>> =
+        runCatching {
+            movieDao.loadAllReservations().toReservations()
+        }
```
어떤 Exception을 막기 위한 runCatching 인가요?_?

murjune replied:
DB 에서 I/O Exception 이 터질 수 있고, 이는 개발자가 예측할 수 없는 영역이기에
비정상 종료를 방지하기 위해 runCatching으로 감싸주었습니다.    

I/O Exception 의 경우 파일 시스템의 문제, 디스크 공간 부족 등으로 발생할 수 있는 것으로 알고 있습니다.  

---  

위와 같은 에러가 발생했을 때, 핸들링하는 로직을 작성해주는 것이 바람직하지만.. 시간에 쫓겨 구현하지 못했습니다.

malibinYun commented on app/src/main/java/woowacourse/movie/data/dao/MovieDao.kt:
```diff
@@ -0,0 +1,47 @@
+    @Insert(onConflict = OnConflictStrategy.IGNORE)
+    fun saveMovieSeats(seats: List<SeatEntity>)
+
+    @Query(
+        """
+        SELECT * FROM 
+        (SELECT * FROM reservation WHERE id = :id) as res
+        JOIN movie_seats AS seat
+        ON res.id = seat.reservationId
+        ORDER BY seat.rowNumber, seat.columnNumber
+    """,
+    )
+    fun loadReservationWithSeats(id: Long): Map<ReservationEntity, List<SeatEntity>>
```
정말 필요한 쿼리인가요?
두 번의 쿼리로 가져왔을 때 성능상의 차이가 컸을까요?

murjune replied:
> 정말 필요한 쿼리인가요?

하나의 트랜잭션으로 처리하는 것이 좋다고 배웠기에 
간단한 쿼리문 이라면 1번에 하는 것을 선호합니다.  

그리고, map 형태로 return 이 가능하다는 것을 공식문서에서 학습하여 위와 같이 작성했습니다!

> 두 번의 쿼리로 가져왔을 때 성능상의 차이가 컸을까요?

성능 차이에 대해서는 잘 모르겠습니다..  
다만, 데이터의 원자성은 보존될 것이라 생각했습니다.

잘 모르는 부분이라서 대답하는게 조심스럽네요..🥲

malibinYun replied:
값을 조회하는 것만 있을 때 트랜젝션은 의미가 있었나요?
트랜젝션은 어떨 때 사용되는 기능일까요?
또, 조회에 있어서 데이터의 원자성 보존이 정말 의미 있는 고려 대상이었나요?

Room에서도 여러 쿼리에 대해 Transaction 기능을 제공한답니다. 그런데 지금 이 쿼리에서 "트렌젝션"과 "데이터의 원자성" 을 논하기에는 뭔가 과해보이네요 🤔

물론 지금 이 영역은 요구사항 범위 바깥의 내용이니 더 깊게 피드백 드리지는 않을게요.

비단 안드로이드 개발자라도, CS영역에 대해 많이 알고 있는 것이 언제나 좋더라구요!
어딘가 주니어들 사이에서는 "안드로이드에도 아키텍처란게 있어요?!" 라는 말을 하기도 하거든요 (실제로 들은말 😇)

murjune replied:
추후, CS에 대해서도 깊이 학습해보곘습니다 💪  
좋은 말씀 감사합니다.

malibinYun commented on app/src/main/java/woowacourse/movie/data/entity/SeatEntity.kt:
```diff
@@ -0,0 +1,31 @@
+import androidx.room.Index
+import androidx.room.PrimaryKey
+
+@Entity(
+    tableName = "movie_seats",
+    foreignKeys = [
+        ForeignKey(
+            entity = ReservationEntity::class,
+            parentColumns = ["id"],
+            childColumns = ["reservationId"],
+            onDelete = ForeignKey.CASCADE,
+        ),
+    ],
+    indices = [Index(value = ["reservationId"])],
+)
```
정말 필요한 외래키 제약인가요?
클라이언트에서의 디비 제약 전략이 개발 생산성과 얼마나 연관성이 있을지 고민해보셔도 좋겠어요.
서버에서 엄격하게 데이터를 다루는 것 만큼 클라이언트에서 엄격히 다루는 것이 정말 개발 리소스를 들일 정도의 것인지도 고민해보세요 :)

murjune replied:
> 정말 필요한 외래키 제약인가요?

여기서 foreignkey를 사용한 이유는 onDelete 옵션을 활용해보기 위해서였습니다.
개발자가 따로 매번 delete 를 명시적으로 작성해주는 것을 잊어버릴 수도 있다고 생각했습니다.  

> 클라이언트에서의 디비 제약 전략이 개발 생산성과 얼마나 연관성이 있을지 고민해보셔도 좋겠어요.

외래키 제약은 조심스러운 부분이라고도 생각되고, 서버에서 데이터를 관리해주는 만큼 
클라이언트에서는 엄격하게 다룰 필요는 굳이 없다고도 생각이 듭니다..

> 개발 생산성에 대해

데이터 생성 및 수정시 항상 무결성 검사하는 비용 혹은 데이터 수정이 까다롭기에 확장성에 문제가 생겨서
그런 것일까요??

---  

아직 클라이언트에서 DB를 다뤄본 경험이 많이 없기에 이와 관련해서 조언 부탁드려도 될까요..?

malibinYun replied:
공부를 목적으로 한 것이라면 말리지는 않겠지만,
미션에서는 여기에 쏟을 시간에 요구사항을 모두 지키는것, 프로젝트를 다듬는 것, 레벨 1에서 배운 것과 여태껏 받아온 피드백들에 대해 지속적으로 고민하면서 프로젝트에 녹여내는 것이 훨씬 더 가치있는 일이라 생각해서요!
요구사항 외에 피드백을 거의 달아두지 않기도 하구요.

---

> 클라이언트에서 DB를 다뤄본 경험이 많이 없기에

백엔드를 공부하시면 됩니다 ㅎㅎ
내 앱과 통신할 내 입맛대로 구성할 서버 구축을 해보세요! 언제까지 앱을 만들기 위해 서버개발자를 구할 수는 없잖아요?

murjune replied:
> 미션에서는 여기에 쏟을 시간에 요구사항을 모두 지키는것, 프로젝트를 다듬는 것, 레벨 1에서 배운 것과 여태껏 받아온 피드백들에 대해 지속적으로 고민하면서 프로젝트에 녹여내는 것이 훨씬 더 가치있는 일이라 생각해서요!  

그동안 배운 것을 리마인드하면서 미션 요구사항을 수행하는 것에 최우선으로 집중하겠습니다!  
좋은 말씀 감사합니다 🙇‍♂️  

> 백엔드를 공부하시면 됩니다 ㅎㅎ
내 앱과 통신할 내 입맛대로 구성할 서버 구축을 해보세요! 언제까지 앱을 만들기 위해 서버개발자를 구할 수는 없잖아요?

추후, 서버 구축하는 경험도 해보도록 하겠습니다!!  
아직 경험치가 적어 보는 시야가 좁은 것 같습니다..😥

malibinYun commented on app/src/main/java/woowacourse/movie/presentation/movieDetail/MovieDetailActivity.kt:
```diff
@@ -0,0 +1,141 @@
+        binding.executePendingBindings()
+    }
+
+    override fun updateDateAdapter(dates: List<String>) {
+        dateAdapter.updateRunningDates(dates)
+    }
+
+    override fun updateTimeAdapter(times: List<String>) {
```
Presenter가 Adapter라는 존재에 대해 이름으로써 관심을 가져야할까요?

murjune replied:
너무 안드로이드 의존성이 결합된 네이밍인 것 같습니다!
수정하겠습니다.

malibinYun commented on app/src/main/java/woowacourse/movie/presentation/reservation/ReservationAdapter.kt:
```diff
@@ -0,0 +1,54 @@
+        return items.size
+    }
+
+    @SuppressLint("NotifyDataSetChanged")
+    fun updateItems(reservations: List<Reservation>) {
+        items = reservations
+        notifyDataSetChanged()
+    }
```
NotifyDataSetChanged는 왜 린트에 잡혔을까요?

murjune replied:
NotifyDataSetChanged 의 경우 호출됐을 때  
oldItems 와 비교하는 로직이 없이 모든 ItemView 들을 다시 그립니다.
변경 사항만 다시 그리는 것에 비해 매우 비효율적이기에 Lint가 잡힙니다.  

따라서, DiffUtil 을 활용하거나 `notifyItemChanged(position)` 를 활용하여 변경된 뷰만 변경되도록 하는 것이   
올바른 로직인 것 같습니다.

malibinYun replied:
DiffUtil을 활용하기 전, 다른 종류의 notify 함수들은 무엇이 있는지 더 알아보세요 :)

malibinYun reviewed (changes requested):
3, 4 단계 미션 고생 많으셨어요.
피드백 반영 후 다시 리뷰 요청 부탁드려요!
화요일에는 다음 미션을 위해 머지하도록 할게요 :)

malibinYun commented:
## 1. Espresso 테스트 에서 dataBinding 객체 가져오기

글쎄요! 저도 잘 모르겠어요! 오둥이가 이슈를 해결해서 저를 알려주시겠어요 ?_?

## 2. Permission 요청 로직

https://stackoverflow.com/questions/74088458/runtime-permission-android-called-twice-if-user-deny-permission
이런 글이 있네요~
이것도 학습해서 저에게 공유해 주시겠어요 ?_?

murjune committed:
:recycle: NotificationDataStore 프로퍼티 네이밍 변경

canNotification -> acceptedPushAlarm

murjune committed:
:sparkles: SettingFragment 완전히 거절 시 App Setting으로 이동

murjune committed:
:recycle: MovieDetailContract 네이밍 변경

- showReleaseDates
-  showReleaseTimes

murjune committed:
:truck: testApplicationContext 으로 네이밍 변경

murjune committed:
:fire: MovieDaoTest repeat(1000) 삭제

murjune committed:
:sparkles: 예매완료 페이지에서 뒤로가기 시 Home 으로 이동

murjune committed:
:recycle: MovieBroadCastReceiver 함수 분리

murjune committed:
:recycle: MovieRepositoryInjector 로 리팩토링

- 중복해서 싱글톤으로 만드는 로직 개선(App 에 있던 로직 삭제)
- MovieRepositoryStore ---> MovieRepositoryInjector 로 네이밍 변경

murjune committed:
:recycle: ktLintFormat

murjune commented:
안녕하세요 말리빈!  
피드백 반영 및 답글 모두 남겨두었습니다.

추가로 Permission 요청 로직 수정했습니다!!
이번에도 잘부탁드립니다.🙇‍♂️  

## SettingFragment Switch ON 눌렀을 때 액션 분기 처리

`shouldShowRequestPermissionRationale` 함수가 처음 거절한 경우와 2번 이상 거절한 경우 모두 true 를 반환하기에
`notificationPreference.hasBeenDeniedPermission` 변수를 추가하여 2번 이상 거절한 경우를 구분하였습니다! 

class 들로 역할 분리하는 것이 좋다고 생각하지만.. 시간이 없어 일단은 SettingFragment 내부에서 분기처리를 하였습니다.

1) API 32 이하인 경우, 권한이 있는 걍우 ---> Switch On 자유로움
```
Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU &&
hasAccessPermission()
```
2) 처음 권한 요청한 경우 ---> 권한 요청
```kotlin
// 1. 권한이 없음?
hasAccessPermission().not() &&
// 2. 2번째 권한 요청?
shouldShowRequestPermissionRationale(POST_NOTIFICATIONS).not() &&
// 3. 이미 거절당한 이력이 없니?
notificationPreference.hasBeenDeniedPermission.not()
```
3) 두 번째 권한 요청한 경우 ---> 권한 요청에 대한 설명 다이얼로그 띄우기
```kotlin
// 해당 함수로 한 번 거절한 이력이 있는지 확인
shouldShowRequestPermissionRationale(POST_NOTIFICATIONS)
```
4) 완전히 거절한 경우 ---> 설정으로 이동
```kotlin
// 1. 권한이 없음?
hasAccessPermission().not() &&
// 2번째 권한 요청?
shouldShowRequestPermissionRationale(POST_NOTIFICATIONS).not() &&
// 3. 이미 거절당한 이력이 있니?
notificationPreference.hasBeenDeniedPermission
```

- 최종 분기처리문
```kotlin
when {
    Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU -> updateAlarmSwitch(true)
    isFirstRequest() -> requestPermissionLauncher.launch(POST_NOTIFICATIONS)
    isSecondRequest() -> {
        notificationPreference.hasBeenDeniedPermission = true
        explanationDialogForPushAlarm.show()
    }
    isCompletelyDenied() -> explanationDialogForNavigateToSetting.show()
    else -> requestPermissionLauncher.launch(POST_NOTIFICATIONS)
}
```


https://github.com/woowacourse/android-movie-theater/assets/87055456/47e8bc76-50c5-4ca9-a23e-03cd759c2cd4

--- 
Espresso 테스트 에서 dataBinding은 아직 해결 못했습니다..  ㅜㅜ

malibinYun commented on app/src/main/java/woowacourse/movie/MovieBroadCastReceiver.kt:
```diff
@@ -0,0 +1,90 @@
+            val reservationId =
+                intent.getLongExtra(PurchaseConfirmationActivity.EXTRA_RESERVATION_ID, -1)
+
+            val notificationDatastore =
+                DefaultNotificationDataStore.instance(context.applicationContext)
+            if (notificationDatastore.acceptedPushAlarm) {
+                sendNotification(context, movieTitle, reservationId)
```
BroadcastReceiver에도 아키텍처를 적용해볼 수 있을까요?

murjune replied:
아키텍쳐를 적용해보라고 말씀하신 것이 살짝 이해가 어렵지만  
현재 `MovieBroadCastReciever` 에서 Notification 를 만들고 매니저에게 알리는 것을 class 분리할 수 있을 것 같습니다!

malibinYun commented on app/src/main/java/woowacourse/movie/MovieReservationApp.kt:
```diff
@@ -0,0 +1,25 @@
+import woowacourse.movie.MovieBroadCastReceiver.Companion.RESERVATION_NOTIFICATION_ACTION
+
+class MovieReservationApp : Application() {
+    private val reservationBroadcastReceiver: MovieBroadCastReceiver by lazy { MovieBroadCastReceiver() }
+
```
동적으로 등록하는 것 말고도, 정적으로 BroadcastReceiver를 등록하는 방법에 대해서도 알아보세요 :)

malibinYun commented on app/src/main/java/woowacourse/movie/presentation/HomeActivity.kt:
```diff
@@ -0,0 +1,93 @@
+        return true
+    }
+
+    private fun isFirstRequest(): Boolean =
+        (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) &&
+            hasAccessPermission().not() &&
+            shouldShowRequestPermissionRationale(POST_NOTIFICATIONS).not() &&
+            notificationDataStore.hasBeenDeniedPermission.not()
```
> shouldShowRequestPermissionRationale 함수가 처음 거절한 경우와 2번 이상 거절한 경우 모두 true 를 반환하기에
notificationPreference.hasBeenDeniedPermission 변수를 추가하여 2번 이상 거절한 경우를 구분하였습니다!

꼭 처음 요청인지가 중요한걸까요?
권한 다이얼로그가 나타났을 때 거절한 여부를 안 것과, 다이얼로그 조차 나타나지 않았는데 거절한 여부를 바로 받을 수 있는 값이 있다면,
해당 값을 가지고 원하는 커스텀 다이얼로그를 띄우면 동작이 매우 이상할까요?
어떤 동작을 구체적으로 원하는지 잘 모르겠네요 🥲

malibinYun commented on app/src/test/java/woowacourse/movie/presentation/movieDetail/MovieDetailPresenterTest.kt:
```diff
@@ -0,0 +1,90 @@
+
+    @Test
+    fun `updateRunMovieTimes를 호출시키면 view의 updateTimeAdapter로 영화 시간 데이터가 전달되는지 확인`() {
+        val expected = cinema.theater.times.map { it.toString() }
+
+        presenter.updateRunMovieTimes()
+
+        verify { view.showReleaseTimes(expected) }
```
테스트 검증부는 하드코딩해보는 건 어떨까요?
관련 좋은 글을 공유드려요 :)
https://jojoldu.tistory.com/615

murjune replied:
코드를 보는 동료 개발자가 한눈에 알아 볼 수 있도록  
하드 코딩하는 방식으로 테스트를 개선해보도록 하겠습니다!

malibinYun commented on app/src/test/java/woowacourse/movie/presentation/movieList/MovieListPresenterTest.kt:
```diff
@@ -0,0 +1,80 @@
+            }
+        }
+
+        private fun generateSeats(): Map<String, Seat> {
+            val seats = mutableMapOf<String, Seat>()
+            val rows = 5
+            val columns = 4
+
+            for (row in 0 until rows) {
+                for (col in 1..columns) {
+                    val seatId = "${('A' + row)}$col"
+                    val grade =
+                        when (row) {
+                            0, 1 -> "B"
+                            2, 3 -> "S"
+                            else -> "A"
+                        }
+                    seats[seatId] = Seat(('A' + row), col, grade)
+                }
+            }
+            return seats
```
테스트를 위한 데이터가 로직으로 되어있다면, 이게 무엇을 담은 데이터인지 테스트를 보는 사람에게 설명서로써의 역할을 해줄 수 있을까요?

murjune replied:
동료 개발자가 이해할 수 있는 더 명시적인 fixture로 개선해보도록 하겠습니다!

malibinYun commented on app/src/test/java/woowacourse/movie/presentation/movieList/cinemaListDialog/ChooseCinemasPresenterTest.kt:
```diff
@@ -0,0 +1,133 @@
+
+        presenter.loadCinema(theaters[0])
+
+        verify { view.showCinemas(expected) }
+    }
```
verify에는 몇 번 불렸는지에 대한 것도 검증할 수 있는 파라미터도 있답니다. 활용해보셔도 좋겠어요!

murjune replied:
exactly 를 활용하여 개선해보겠습니다!

malibinYun reviewed (approved):
2주차 미션 고생 많으셨어요 :)
다음 미션에 집중하기 위해 이번 미션은 이만 머지하도록 할게요. 
남은 미션들도 잘 수행하시길 응원할게요! 궁금한 게 있다면 언제든지 DM 주셔도 좋아요 :)
